src/blockrelay/thinblock.cpp

```diff
@@ -484,40 +478,37 @@ bool CXRequestThinBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
         }
     }
 
+    std::vector<CTransaction> vTx;
+    CBlockIndex *hdr = LookupBlockIndex(inv.hash);
+    if (!hdr)
     {
-        LOCK(cs_main);
-        std::vector<CTransaction> vTx;
-        BlockMap::iterator mi = mapBlockIndex.find(inv.hash);
-        if (mi == mapBlockIndex.end())
+        dosMan.Misbehaving(pfrom, 20);
+        return error("Requested block is not available");
+    }
+    else
+    {
+        CBlock block;
+        const Consensus::Params &consensusParams = Params().GetConsensus();
+        if (!ReadBlockFromDisk(block, hdr, consensusParams))
         {
```

Comment:
I believe we still need cs_main for reading/writing blocks to disk.  Level db is not really thread safe from my experience with the utxo.  This is another area we could apply a different criticalsection for doing these reads/writes.

Reply:
see https://github.com/google/leveldb/blob/master/doc/index.md#concurrency.
Something's not thread safe in the utxo then... maybe bitcoind's cache?

But also most people are still using the "flat" block access which should be thread safe because written files are immutable.  So if blockdb turns out to be not thread safe or leveldb buggy we should lock it at a lower level.  

(also, I've been running 2 nodes one on each fork for a few days now with no issues)